feat: make GraphQL a peer dependency, support GraphQL v15.0

JohnDaly opened the pull request:
The `graphql` package throws errors if it's not treated as a [singleton dependency](https://github.com/graphql/graphql-js/blob/94234c8ce0a33cd8d884d0fced3d7d107db8f8aa/src/jsutils/instanceOf.js#L32). It's recommended for libraries to install it as a peer dependency (see: https://github.com/graphql/graphql-js/issues/594#issuecomment-1189850052).

To allow MSW to work with projects that are using `graphql` v15, I've made a few small changes to the code to avoid using `OperationTypeNode` as an enum. The updated code is compatible with both v15 and v16 of the `graphql` package.

JohnDaly committed:
feat: move 'graphql' to peerDependencies. Add support for v15 of 'graphql'

mattcosta7 commented on package.json:
```diff
@@ -134,6 +134,7 @@
     "webpack-dev-server": "^3.11.2"
   },
   "peerDependencies": {
+    "graphql": "^15.0.0 || ^16.0.0",
```
can graphql be an optional peer?  I'm curious if it's required for users who only define `rest` methods?

JohnDaly replied:
That would be ideal, however it would require additional changes. Currently, the `graphql` package is included in the main entrypoint, so it's a required peer dependency of `msw`.

kettanaito replied:
I believe `graphql` should come in as an optional dependency. Otherwise, we'd be demanding our users who are not using GraphQL in their apps to install `graphql` alongside MSW. 

By making `graphql` an optional peer dependency (similar to how we list `typescript`), we are not making any breaking changes to our consumers, actually:

- Devs who are using GraphQL in their apps, already have GraphQL installed—no change here for them;
- Devs who are not using GraphQL are not affected at all.

JohnDaly replied:
@kettanaito do you have an idea of how we might be able to make `graphql` optional? 

As far as I can tell, the only code being imported from `graphql` that is needed at runtime by `msw` is the `parse` function. It is used [here](https://github.com/mswjs/msw/blob/fc7f00c96e79ad0be83cbaf7c82ceff23f0c4355/src/utils/internal/parseGraphQLRequest.ts#L44).

My thought was to do something like this, to treat the package as optional:

```ts
// Since 'graphql' is an optional peer dependency,
// we'll attempt to import it dynamically
let parse: typeof import('graphql').parse
try {
  parse = require('graphql').parse
} catch (e) {}
```

What I'm unsure of is if the ESM output of `msw` will work with this approach.

The other option I considered is to move the code that depends on `graphql` to a different entrypoint (e.g. `msw/graphql`). This would be a breaking change though.

What do you think?

kettanaito replied:
If we have proper code splitting in place, if you're not using `graphql` from `msw`, your app shouldn't get the line of code where we do `import { parse } from 'graphql'`. This is something I'd kindly ask you to test out (steps described in some other comment on this pull request). 

And in terms of package.json, I believe you can take a look how we specify `"typescript"`, as it's an optional peer dependency we already have. 

> Side note: I think NPM also force-installs optional peer dependencies, which is sort of okay. As long as we don't break users' apps by forcing them to manually install a dependency they don't otherwise need.

JohnDaly replied:
I'm not sure we can solve this with code splitting, without introducing a breaking change.

The problem is that the `graphql` import occurs within the main entrypoint of `msw`. When a consumer imports `msw`, and they don't have `graphql` in their `node_modules`, the `msw` entrypoint will fail to load because there will be a `MODULE_NOT_FOUND` error when trying to resolve `graphql`.

To code split, we would need to move the `graphql` import to a different entrypoint, which would be a breaking change.

I've opted to do the following in https://github.com/mswjs/msw/pull/1356/commits/7a8482ab8a1745533d9b845c0a0e6713eb79c2d1:
1. Mark `graphql` as an optional peer dependency
2. Dynamically load `graphql` within the `parseQuery` function using `require`. This function should only be called when the consumer is attempting to mock GraphQL with `msw` 
3. Update all static imports of `graphql` in the project to be `type imports`

I tested these changes by using my version of `msw` with the [with-msw](https://github.com/vercel/next.js/tree/canary/examples/with-msw) Next.js example. I was able to run the app, and mock rest requests, without needing to install `graphql`.

The CodeSandbox example linked to this PR now compiles. It does warn about the `graphql` module not being found, but this is just a console warning and doesn't affect the application's ability to mock rest requests.

kettanaito replied:
I agree with your approach, I think it's great! Thanks for looking into this 👍

kettanaito commented on package.json:
```diff
@@ -134,6 +134,7 @@
     "webpack-dev-server": "^3.11.2"
   },
   "peerDependencies": {
+    "graphql": "^15.0.0 || ^16.0.0",
```
But one concern I do have is to double-check MSW's tree shaking. 

1. List this as an optional peer dependency.
2. Create an empty project and install MSW there.
3. Don't use `graphql`. Would your app compile?

kettanaito replied:
This is what you've confirmed with the sandbox, right, @JohnDaly?

JohnDaly replied:
Yes, I was able to run the Sandbox and use MSW to mock REST requests

JohnDaly committed:
feat: marking 'graphql' as an optional peer dependency

The 'graphql' code that is needed at runtime will be dynamically loaded. Existing static imports of
'graphql' are designated as type imports.

kettanaito commented on package.json:
```diff
@@ -116,6 +115,7 @@
     "eslint-plugin-prettier": "^3.4.0",
     "fs-extra": "^10.0.0",
     "fs-teardown": "^0.3.0",
+    "graphql": "^16.3.0",
```
Is it safe to migrate to v16 given the optional range is 15/16? Are there any features that are dropped in v16 that we may accidentally rely upon?

JohnDaly replied:
Since v16 is the version that MSW has used in the latest release, I think it makes sense to keep it as the devDependency version.

There are two main changes in this PR that affect run-time compatibility:
1. Changing usage of `OperationTypeNode` from enum to string
2. The `parse` function

Changing `OperationTypeNode` should be safe to do, the `enum` in v16 maps to the same strings that were used in v15.

The `parse` function from `graphql` is the only piece of runtime code that is actually used by MSW. It's being used to get the `OperationDefinitionNode` for the GraphQL operation that is intercepted. From what I can tell, the fields that MSW is interested in on `OperationDefinitionNode` are unchanged between v15 and v16.

Based on this, I think it should be safe to list v15 as a dependency. 

It's also worth noting that v15 is, at the time of writing, the most commonly downloaded version of `graphql` on NPM (3+ million downloads a week, compared to ~2 million for v16). It would be good to be backwards compatible with v15, because lots of MSW users probably have that version in their projects.
